### PR TITLE
Implement WeeklyArenaState query

### LINKED ISSUE #252: Add query for WeeklyArenaState

### PR DESCRIPTION
**Pull request by ipdae**
resolve #252 

```
            query {
                stateQuery {
                    weeklyArena(index: 0) {
                        address
                        ended
                        orderedArenaInfos {
                            agentAddress
                            avatarAddress
                            arenaRecord {
                                win
                                lose
                                draw
                            }
                            level
                            combatPoint
                            armorId
                            active
                            dailyChallengeCount
                            score
                        }
                    }
                }
            }
```

### CODE REVIEW

File: NineChronicles.Headless/GraphTypes/StateQuery.cs [hunk 1]
```diff
@@ -42,6 +42,20 @@ public StateQuery()
             Field<ShopStateType>(
                 name: "shop",
                 resolve: context => new ShopState((Dictionary) context.Source.GetState(Addresses.Shop)));
+            Field<WeeklyArenaStateType>(
+                name: "weeklyArena",
+                arguments: new QueryArguments(
+                    new QueryArgument<NonNullGraphType<IntGraphType>>
+                    {
+
```
Comment (moreal) [inline NineChronicles.Headless/GraphTypes/StateQuery.cs:50]: This line needs to be removed. 👀

File: NineChronicles.Headless.Tests/GraphTypes/StandaloneQueryTest.cs [hunk 2]
```diff
@@ -820,6 +820,104 @@ public async Task StateQueryShopState(ItemType itemType)
...
+                                    ["dailyChallengeCount"] = 5,
+                                    ["score"] = 1000,
+                                }
+                            },
+                        },
+                    },
+                },
+                queryResult.Data
+            );
+        }
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/StandaloneQueryTest.cs:919]: I want to suggest to test in more smaller scope. It seems to test `SetAvatarState` action also. So I suggest to make a new test class like `WeeklyArenaStateTest`. You can see `CollectionMapTypeTest` as an example.

https://github.com/planetarium/NineChronicles.Headless/blob/6e4b0892624ec007fced271b9785047257bae0d0/NineChronicles.Headless.Tests/GraphTypes/States/Models/CollectionMapTypeTest.cs#L23-L25

Please leave opinions if you have! Or if you agreed with my suggestion but you are hard to acheive it, please leave any question also, don't feel hard 😀

Comment (ipdae) [reply NineChronicles.Headless.Tests/GraphTypes/StandaloneQueryTest.cs:919]: My intention was check cases when `OrderdArenaInfos` is empty and has a value. i'll try simplify the test by using `Theory`.

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopStateTypeTest.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                new Dictionary<string, object>
+                {
+                    ["itemId"] = new Guid("d3d9ac06-eb91-4cc4-863a-5b4769ad633e"),
+                    ["itemType"] = ItemType.Costume.ToString().ToUpper(),
+                    ["itemSubType"] = "FULL_COSTUME",
+                },
+            },
+        };
+    }
+}
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopStateTypeTest.cs:180]: IMO I believe these tests can have more small scope without `action` and `blockchain` and their names can be more expressionable without `StateQuery` prefix.

I want to suggest to test with [`GraphQLTestUtils`](https://github.com/planetarium/NineChronicles.Headless/blob/738d21d23c3a895e40f3bbd6ef57fd0f44e125e2/NineChronicles.Headless.Tests/GraphQLTestUtils.cs#L10-L28), not `GraphQLTestBase`. The static class helps to execute query with a `IObjectGraphType` as root query easily. Then it can be free from `BlockChain<T>` and be away from not related graph types in implementation (e.g. `StateQuery`), also the `StateQuery` prefix on all tests can be removed. 😀 

Of course, I think this opinion can be wrong anytime because I didn't understand actions and states of `Lib9c` well yet. You can leave a comment in free. And you can also with call (e.g. Slack) if you want to ask or discuss this.

File: NineChronicles.Headless/GraphTypes/States/ArenaRecordType.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        {
+            Field<IntGraphType>(
+                nameof(ArenaInfo.Record.Win),
+                resolve: context => context.Source.Win);
+            Field<IntGraphType>(
+                nameof(ArenaInfo.Record.Lose),
+                resolve: context => context.Source.Lose);
+            Field<IntGraphType>(
+                nameof(ArenaInfo.Record.Draw),
+                resolve: context => context.Source.Draw);
```
Comment (moreal) [inline NineChronicles.Headless/GraphTypes/States/ArenaRecordType.cs:18]: The fields seem not nullable though it can be 0.

https://github.com/planetarium/lib9c/blob/a5339c0de85ec6cd9523e355f97096487eae1acf/Lib9c/Model/State/WeeklyArenaState.cs#L317-L319

```suggestion
            Field<NonNullGraphType<IntGraphType>>(
                nameof(ArenaInfo.Record.Win),
                resolve: context => context.Source.Win);
            Field<NonNullGraphType<IntGraphType>>(
                nameof(ArenaInfo.Record.Lose),
                resolve: context => context.Source.Lose);
            Field<NonNullGraphType<IntGraphType>>(
                nameof(ArenaInfo.Record.Draw),
                resolve: context => context.Source.Draw);
```

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingMapStateTypeTest.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        public RankingMapStateTypeTest(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Theory]
+        [MemberData(nameof(Members))]
+        public async Task QueryRankingMapState(bool rankingInfoExist, List<object> expected)
+        {
+            var userPrivateKey =
+                new PrivateKey(ByteUtil.ParseHex("b934cb79757b1dec9f89caa01c4b791a6de6937dbecdc102fbdca217156cc2f5"));
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingMapStateTypeTest.cs:25]: It seems used in multiple tests. How about making fixture containing such them?

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/AvatarStateTypeTest.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+using Xunit;
+using static NineChronicles.Headless.Tests.GraphQLTestUtils;
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class AvatarStateTypeTest
+    {
+        [Theory]
+        [MemberData(nameof(Members))]
+        public async Task QueryAvatarState(AvatarState avatarState, Dictionary<string, object> expected)
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/AvatarStateTypeTest.cs:14]:
```suggestion
        public async Task Query(AvatarState avatarState, Dictionary<string, object> expected)
```

Isn't it enough with `Query` as a test name, like you used in `ArenaInfoTypeTest`?

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingInfoTypeTest.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+using Xunit;
+using static NineChronicles.Headless.Tests.GraphQLTestUtils;
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class RankingInfoTypeTest
+    {
+        [Theory]
+        [MemberData(nameof(Members))]
+        public async Task QueryRankingInfo(RankingInfo rankingInfo, object expected)
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingInfoTypeTest.cs:14]:
```suggestion
        public async Task Query(RankingInfo rankingInfo, object expected)
```

Isn't it enough with `Query` as a test name, like you used in `ArenaInfoTypeTest`?

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingMapStateTypeTest.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+using Xunit;
+using static NineChronicles.Headless.Tests.GraphQLTestUtils;
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class RankingMapStateTypeTest
+    {
+        [Theory]
+        [MemberData(nameof(Members))]
+        public async Task QueryRankingMapState(RankingMapState rankingMapState, object expected)
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/RankingMapStateTypeTest.cs:14]:
```suggestion
        public async Task QueryRankingMapState rankingMapState, object expected)
```

Isn't it enough with `Query` as a test name, like you used in `ArenaInfoTypeTest`?

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopItemTypeTest.cs [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+using Xunit;
+using static NineChronicles.Headless.Tests.GraphQLTestUtils;
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class ShopItemTypeTest
+    {
+        [Theory]
+        [MemberData(nameof(Members))]
+        public async Task QueryShopItem(
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopItemTypeTest.cs:16]:
```suggestion
        public async Task Query(
```

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopStateTypeTest.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+using NineChronicles.Headless.GraphTypes.States;
+using Xunit;
+using static NineChronicles.Headless.Tests.GraphQLTestUtils;
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class ShopStateTypeTest
+    {
+        [Fact]
+        public async Task QueryShopState()
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/ShopStateTypeTest.cs:14]:
```suggestion
        public async Task Query()
```

File: NineChronicles.Headless.Tests/GraphTypes/States/Models/WeeklyArenaStateTypeTest.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+
+namespace NineChronicles.Headless.Tests.GraphTypes.States.Models
+{
+    public class WeeklyArenaStateTypeTest
+    {
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public async Task QueryWeeklyArenaState(bool ended)
```
Comment (moreal) [inline NineChronicles.Headless.Tests/GraphTypes/States/Models/WeeklyArenaStateTypeTest.cs:16]:
```suggestion
        public async Task Query(bool ended)
```

File: NineChronicles.Headless.Tests/SetAvatarState.cs [hunk 12]
```diff
@@ -59,7 +59,8 @@ public override IAccountStateDelta Execute(IActionContext context)
```
Comment (moreal) [inline NineChronicles.Headless.Tests/SetAvatarState.cs:59]: The changes of tests in this pull request became not related to `BlockChain<T>`. So it seems not related to also `SetAvatarState` action to me, doesn't it?

**Review by moreal — APPROVED** [review]
👍

**Review by area363 — APPROVED** [review]
